envoy/config/xds_config_tracker.h

```diff
@@ -0,0 +1,115 @@
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include "envoy/common/optref.h"
+#include "envoy/config/subscription.h"
+#include "envoy/config/typed_config.h"
+#include "envoy/protobuf/message_validator.h"
+
+#include "source/common/protobuf/protobuf.h"
+
+#include "google/rpc/status.pb.h"
+
+namespace Envoy {
+namespace Config {
+
+// The status of processing Delta/DiscoveryResponse.
+enum ProcessingState {
+  // Successfully got the resources or message.
+  RECEIVED = 0,
+  // Failed to apply the resources.
+  FAILED = 1,
+};
+
+struct ProcessingDetails {
+  ProcessingDetails(const ProcessingState state) : state_(state){};
+  ProcessingDetails(const ProcessingState state, const ::google::rpc::Status error_detail)
+      : state_(state), error_detail_(error_detail){};
+  const ProcessingState state_;
+  ::google::rpc::Status error_detail_;
+};
+
+/**
+ * An interface for hooking into xDS update events to provide the ablility to use some external
+ * processor in xDS update. This traker provides the process point when the discovery response
+ * is received, when the resoures are successfully processed and applied, and when there is any
+ * failure.
+ *
+ * Instance of this interface get invoked on the main Envoy thread. Thus, it is important
+ * for implementations of this interface to not execute any blocking operations on the same
+ * thread.
+ */
+class XdsConfigTracker {
+public:
+  virtual ~XdsConfigTracker() = default;
+
+  /**
+   * Invoked when SotW xDS configuration updates have been successfully parsed, applied on
+   * the Envoy instance, and are about to be ACK'ed.
+   *
+   * @param type_url The type url of xDS message.
+   * @param resources A list of decoded resources to add to the current state.
+   */
+  virtual void onConfigIngested(const absl::string_view type_url,
+                                const std::vector<DecodedResourcePtr>& resources) PURE;
+
+  /**
+   * Invoked when Delta xDS configuration updates have been successfully parsed, applied on
+   * the Envoy instance, and are about to be ACK'ed.
+   *
+   * @param type_url The type url of xDS message.
+   * @param added_resources A list of decoded resources to add to the current state.
+   * @param removed_resources A list of resources to remove from the current state.
+   */
+  virtual void
+  onConfigIngested(const absl::string_view type_url,
+                   const std::vector<DecodedResourcePtr>& added_resources,
+                   const Protobuf::RepeatedPtrField<std::string>& removed_resources) PURE;
+
+  /**
+   * Invoked when received or failed to process the DiscoveryResponse.
```

Comment:
This is a bit of a confusing name / semantics. First, what is a "failed" `DiscoveryResponse`? Second, how do we think about this lifecycle-wise vs. `onConfigIngested`? Always first right?

An xDS config tracker is tracking the applied configuration, not the transport protocol or proxy NACKs due to rejections I think. Or does it do this as well? I think being clear on the mission here, and whether we are tracking the applied configuration, vs. the wire configuration, is really important to a clean interface.

Reply:
@adisuissa as well

Reply:
We tracked the ingested configurations, and also provide the hooks for the message level when receiving or there is any exception during processing the response.

Reply:
Thanks for reviewing!

This hook is like we will throw an exception when we fail to decode and apply the configs from `DiscoveryResponse`, the hook is applied there and when receiving the message as well. I am thinking we can change the name to `onConfigUpdateReceivedorFailed`. WDYT? Thanks!

Reply:
I think it might be better to split this into two separate callbacks for the two cases. That way we have less need for each implementation to switch themselves on the two fairly distinct cases. Also, I still think the different cases are very confusing name wise. Here would be my suggestion, three methods:
* `onConfigReceived` which is called when config is first received in the xDS ingestion lifecycle (I think?! please clarify when this is and why we even need to think about this).
* `onConfigRejected` which is called if config is rejected during xDS ingestion.
* `onConfigAccepted` which is called if config is accepted during xDS ingestion (i.e. what `onConfigIngested` is today).

Reply:
+1 on using explicit callback methods rather than sending the state.

Feel free to correct me if I'm wrong, but compared to the current state of the protocol, does this tracker essentially adds the ability to know when a resource was received, or is it something else?
In other words, at the moment, given a response, Envoy processes it and either accepts (all resources "ingested") or rejects the response. This tracker allows to know when a resource arrived (received) at the Envoy.
Can you provide more details around the usefulness of this new info?

Reply:
Another thought of where things may be different is if the Envoy received a resource that it hasn't subscribed for, and so you'll see it on the "received" side, but not on the processed?

Reply:
Thank you all for the input. I am aligned with the `onConfigReceived` may not add too much value here. We care more about the accepted configs and any exceptions, and if there is any error or unsubscribed type_url, we either directly NACK or throw an exception.